language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -77,45 +73,117 @@ import           Data.HashMap.Monoidal
 prerequisite. The CEK machine correctly handles name shadowing.
 -}
 
+type TermWithMem uni = WithMemory Term uni
+type TypeWithMem uni = Type TyName uni ExMemory
+type KindWithMem = Kind ExMemory
+
+-- 'Values' for the modified CEK machine.
+data CekVal uni =
+    -- TODO: we probably want to store a @Some (ValueOf uni)@ here, but then we have trouble in
+    -- 'readKnownCek'. I'll reconsider the way we deal with annotations once again.
+    VCon (TermWithMem uni)
+  | VTyAbs ExMemory TyName KindWithMem (TermWithMem uni) (CekValEnv uni)
+  | VLamAbs ExMemory Name (TypeWithMem uni) (TermWithMem uni) (CekValEnv uni)
+  | VIWrap ExMemory (TypeWithMem uni) (TypeWithMem uni) (CekVal uni)
+  | VBuiltin
+      ExMemory
+      (CekValEnv uni)   -- Initial environment, used for evaluating every argument
+      StagedBuiltinName
+      Int               -- Number of arguments to be provided
+      [TypeWithMem uni] -- The types the builtin is to be instantiated at.
+                        -- We need these to construct a term if the machine is returning a stuck partial application.
+      [CekVal uni]      -- Arguments we've computed so far.
+    deriving (Show, Eq) -- Eq is just for tests.
+
+instance (Closed uni, GShow uni, uni `Everywhere` PrettyConst) => PrettyBy PrettyConfigPlc (CekVal uni) where
+    prettyBy cfg = prettyBy cfg . dischargeCekVal
+
+type instance UniOf (CekVal uni) = uni
+
+instance (Closed uni, uni `Everywhere` ExMemoryUsage) => FromConstant (CekVal uni) where
+    fromConstant = VCon . fromConstant
+
+instance AsConstant (CekVal uni) where
+    asConstant (VCon term) = asConstant term
+    asConstant _           = Nothing
+
+instance ToExMemory (CekVal uni) where
+    toExMemory = \case
+        VCon t -> termAnn t
+        VTyAbs ex _ _ _ _ -> ex
+        VLamAbs ex _ _ _ _ -> ex
+        VIWrap ex _ _ _ -> ex
+        VBuiltin ex _ _ _ _ _ -> ex
+
+type CekValEnv uni = UniqueMap TermUnique (CekVal uni)
+
+-- See Note [Scoping].
+-- | Instantiate all the free variables of a term by looking them up in an environment.
+dischargeCekValEnv :: CekValEnv uni -> TermWithMem uni -> TermWithMem uni
+dischargeCekValEnv valEnv =
+    -- We recursively discharge the environments of Cek values, but we will gradually end up doing
+    -- this to terms which have no free variables remaining, at which point we won't call this
+    -- substitution function any more and so we will terminate.
+    termSubstFreeNames $ \name -> do
+        val <- lookupName name valEnv
+        Just $ dischargeCekVal val
+
+unstageBuiltinName :: ann -> StagedBuiltinName -> Builtin ann
+unstageBuiltinName ann (StaticStagedBuiltinName name)  = BuiltinName ann name
+unstageBuiltinName ann (DynamicStagedBuiltinName name) = DynBuiltinName ann name
+
+mkBuiltinApp :: ExMemory -> Builtin ExMemory -> [TypeWithMem uni] -> [TermWithMem uni] -> TermWithMem uni
+mkBuiltinApp ex bn tys args = mkIterApp ex (mkIterInst ex (Builtin ex bn) tys) args
```

Comment:
So currently we just assume that type and term arguments are not interleaved, right? It's worth writing this assumption down somewhere (unless it's already there and I just fail to see it).

Reply:
I'll stick a comment in.

Reply:
Oh, that's a bit nasty.  I'm trying to ensure that we deal properly with interleaved type/value arguments by handling arities as `TermArg | TypeArg` lists.  The frame accumulates types and values separately and passes the terms to the CAM when we're fully applied.  In order to convert a frame into a correct application we're going to have to retrieve the arity and use it to interleave the term and type arguments properly.  We only have to do this for errors and when then machine returns a partially-applied builtin, so maybe that's OK.  The alternative is presumably to accumulate  list of `Type + Value` values and then if we have a successful application, to filter out the values and pass them to the CAM, which is unpleasantly inefficient. I'll go ahead and write the re-interleaving function.

Reply:
I wanted to say "just don't bother with this, we're going to move this all to Untyped Plutus Core and there we don't have types", but even though we don't have types, we still need to correctly put `force` in place of a type application.

So yeah, please bother with this.